src/signature.js

```diff
@@ -50,250 +63,272 @@ const spellNumber = x => ({
 
-//-----------------------------------------------------------------------------
+// ----------------------------------------------------------------------------
 //
 // Type classes
 //
-//-----------------------------------------------------------------------------
+// ----------------------------------------------------------------------------
 
-// TypeClassMap -> String -> TypeClass
-const lookupTypeClass = tcm => (name) => {
+// lookupTypeClass :: TypeClassMap -> String -> TypeClass
+const lookupTypeClass = tcm => name => {
   const tc = tcm[name];
   if (!tc) {
-    const allTypeClasses = R.keys(tcm).join(', ');
-    throw new TypeError(
-      `Type class ${name} not found. ` +
-      `Available type classes are: ${allTypeClasses}`,
-    );
+    const allTypeClasses = S.pipe ([S.keys, S.joinWith (', ')]) (tcm);
+    throw new TypeError (`Type class ${name} not found. Available type `
+      + `classes are: ${allTypeClasses}`);
   }
 
   return tc;
 };
 
-// [SignatureConstraint] -> StrMap String
-export const constraintNames = R.converge(R.zipObj, [
-  R.pluck('typevar'),
-  R.pluck('typeclass'),
-]);
-
-// TypeClassMap -> [SignatureConstraint] -> StrMap [TypeClass]
-export const constraints = uncurry2(
-  tcm => R.compose(
-    R.map(R.of),
-    R.map(lookupTypeClass(tcm)),
-    constraintNames,
-  ),
-);
-
-// :: [TypeClass] -> TypeClassMap
-const indexTypeClasses = R.indexBy(R.compose(
+// indexTypeClasses :: Array TypeClass -> TypeClassMap
+const indexTypeClasses = memoize (indexBy (S.pipe ([
+  S.prop ('name'),
   stripNamespace,
-  R.prop('name'),
-));
+])));
 
-//-----------------------------------------------------------------------------
+// ----------------------------------------------------------------------------
 //
 // Types
 //
-//-----------------------------------------------------------------------------
-
-// :: { children :: [a] } -> [a]
-const children = R.prop('children');
-
-// :: { children :: [a] } -> a
-const firstChild = R.compose(R.prop(0), children);
-
-// :: Object -> String -> Boolean
-const typeEq = R.propEq('type');
-
-// :: SignatureEntry -> Boolean
-const hasChildren = R.compose(R.not, R.isEmpty, R.prop('children'));
-
-//  :: Type -> (Type -> Type)
-const fromUnaryType = t => $.UnaryType(
-  t.name,
-  t.url,
-  t._test, // eslint-disable-line no-underscore-dangle
-  t.types.$1.extractor
-);
-//  :: Type -> (Type -> Type -> Type)
-const fromBinaryType = t => $.BinaryType(
-  t.name,
-  t.url,
-  t._test, // eslint-disable-line no-underscore-dangle
-  t.types.$1.extractor,
-  t.types.$2.extractor
-);
-
-// :: (Type, [Type]) -> ()
-const checkTypeArity = (type, argTypes) => {
+// ----------------------------------------------------------------------------
+
+// children :: { children :: Array a } -> Array a
+const children = S.prop ('children');
+
+// firstChild :: { children :: Array a } -> a
+const firstChild = S.pipe ([
+  children,
+  S.head,
+  unchecked.fromMaybe (undefined),
```

Comment:
Why return `undefined` rather than `S.Nothing`?

Reply:
Actually, this function will never receive an object that has an empty `children` property, so we can go even simpler and do with `xs => xs[0]`.

Reply:
Sounds good. You should update the type signature accordingly:

```diff
-// firstChild :: { children :: Array a } -> a
+// firstChild :: { children :: NonEmpty (Array a) } -> a
```

I think this is the clearest implementation:

```javascript
const firstChild = r => r.children[0];
```

I also suggest aligning the type signature with the identifier on the subsequent line:

```javascript
//    firstChild :: { children :: NonEmpty (Array a) } -> a
const firstChild = r => r.children[0];
```